src/modules/acbs/acbs-authentication.service.ts

```diff
@@ -0,0 +1,94 @@
+  private static readonly sessionIdCookieName = 'JSESSION';
+  private static readonly sessionIdCookieSeparator = ';';
+
+  constructor(
+    @Inject(AcbsConfig.KEY)
+    private readonly config: Pick<ConfigType<typeof AcbsConfig>, 'apiKey' | 'apiKeyHeaderName' | 'authentication'>,
+    private readonly httpService: HttpService,
+  ) {}
+
+  async getIdToken(): Promise<string> {
+    const sessionId = await this.createSession();
+    return this.getIdTokenForSession(sessionId);
+  }
+
+  private async createSession(): Promise<string> {
+    const createSessionResponse = await lastValueFrom(
+      this.httpService
+        .post<never>(
+          AcbsAuthenticationService.sessionsPath,
+          { loginName: this.config.authentication.loginName, password: this.config.authentication.password },
+          {
+            baseURL: this.config.authentication.baseUrl,
+            headers: {
+              'Content-Type': 'application/json',
+              [this.config.apiKeyHeaderName]: this.config.apiKey,
+            },
+          },
+        )
+        .pipe(
+          catchError((error: Error) => {
+            throw new AcbsAuthenticationFailedException('Failed to create a session with the IdP.', error);
+          }),
+        ),
+    );
+    return this.extractSessionIdFromCreateSessionResponse(createSessionResponse);
+  }
+
+  private extractSessionIdFromCreateSessionResponse(response: AxiosResponse): string {
+    const sessionIdCookie = response.headers['set-cookie'].find((cookie) => cookie.startsWith(AcbsAuthenticationService.sessionIdCookieName));
+
+    if (!sessionIdCookie) {
+      throw new AcbsAuthenticationFailedException('Session cookie was not returned by the IdP.');
+    }
+
+    return sessionIdCookie.includes(AcbsAuthenticationService.sessionIdCookieSeparator)
+      ? sessionIdCookie.substring(0, sessionIdCookie.indexOf(AcbsAuthenticationService.sessionIdCookieSeparator))
+      : sessionIdCookie;
+  }
+
+  private async getIdTokenForSession(sessionId: string): Promise<string> {
+    const connectResponse = await lastValueFrom(
+      this.httpService
+        .get<IdpConnectResponse>(AcbsAuthenticationService.connectPath, {
+          baseURL: this.config.authentication.baseUrl,
+          params: { client_id: this.config.authentication.clientId },
+          headers: {
+            'Content-Type': 'application/x-www-form-urlencoded',
+            [this.config.apiKeyHeaderName]: this.config.apiKey,
+            Cookie: sessionId,
+          },
+        })
+        .pipe(
+          catchError((error: Error) => {
+            throw new AcbsAuthenticationFailedException('Failed to get a token from the IdP.', error);
+          }),
+        ),
+    );
+
+    return this.extractIdTokenFromConnectResponse(connectResponse);
+  }
+
+  private extractIdTokenFromConnectResponse(response: AxiosResponse<IdpConnectResponse>): string {
+    const idToken = response.data?.id_token;
```

Comment:
We can simplify this into
```js
const { id_token : idToken } = response.data;
```

Reply:
Thanks Gabriel, I've gone for
```javascript
const { id_token: idToken } = response.data || {};
```
in case `response.data` is null or undefined

Reply:
this may not work with the if statement below. 
```js
    if (!idToken) {
      throw new AcbsAuthenticationFailedException('ID token was not returned by the IdP.');
    }
```
this will not be triggered if idToken = {}

Reply:
The above code sets `idToken` to `{}.id_token` (which is `undefined`) if `response.data` is `undefined` because the expression on the right-hand side is evaluated before the object destructuring

Reply:
ok cool